clang/test/SemaCXX/cheri/cheri-capability-qualifier-declspec.cpp

```diff
@@ -0,0 +1,17 @@
+// RUN: %cheri_cc1 -o - %s -fsyntax-only -verify
+// expected-no-diagnostics
+
+void * __capability test_decltype(void *p) {
+  decltype(p) __capability c = (__cheri_tocap decltype(p) __capability)p;
+  return c;
+}
+
+// TODO: Clang can't currently represent this and so falls back on
+// "__capability only applies to pointers; type here is 'T'". Ideally we would
+// support this.
+//template<typename T>
```

Comment:
I'd prefer if we uncommented this to test the current behaviour and add a FIXME: comment.

Reply:
I assume making this work would involve delaying the handling of __capability until template instantiation and probably requires quite a few more changes. Probably not particularly important to fix considering that you can use `T *p` instead.

Reply:
Yes, exactly, it's just annoying (and you also potentially need to specialise for references and other such things..), unlike the others which can't be worked around.